subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java

```diff
@@ -40,39 +41,47 @@
     private static final Logger LOGGER = Logging.getLogger(SkipEmptySourceFilesTaskExecuter.class);
     private final TaskInputsListener taskInputsListener;
     private final BuildOutputCleanupRegistry buildOutputCleanupRegistry;
+    private final TaskOutputsGenerationListener taskOutputsGenerationListener;
     private final TaskExecuter executer;
 
-    public SkipEmptySourceFilesTaskExecuter(TaskInputsListener taskInputsListener, BuildOutputCleanupRegistry buildOutputCleanupRegistry, TaskExecuter executer) {
+    public SkipEmptySourceFilesTaskExecuter(TaskInputsListener taskInputsListener, BuildOutputCleanupRegistry buildOutputCleanupRegistry, TaskOutputsGenerationListener taskOutputsGenerationListener, TaskExecuter executer) {
         this.taskInputsListener = taskInputsListener;
         this.buildOutputCleanupRegistry = buildOutputCleanupRegistry;
+        this.taskOutputsGenerationListener = taskOutputsGenerationListener;
         this.executer = executer;
     }
 
     public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionContext context) {
         FileCollection sourceFiles = task.getInputs().getSourceFiles();
         if (task.getInputs().getHasSourceFiles() && sourceFiles.isEmpty()) {
             TaskArtifactState taskArtifactState = context.getTaskArtifactState();
-            FileCollection outputFiles = taskArtifactState.getExecutionHistory().getOutputFiles();
+            TaskExecutionHistory executionHistory = taskArtifactState.getExecutionHistory();
+            Set<File> outputFiles = executionHistory.getPreviousOutputs();
             if (outputFiles == null) {
                 state.setOutcome(TaskExecutionOutcome.NO_SOURCE);
                 LOGGER.info("Skipping {} as it has no source files and no history of previous output files.", task);
             } else if (outputFiles.isEmpty()) {
                 state.setOutcome(TaskExecutionOutcome.NO_SOURCE);
                 LOGGER.info("Skipping {} as it has no source files and no previous output files.", task);
             } else {
-                Set<File> outputFileSet = outputFiles.getFiles();
+                boolean cleanupDirectories = true;
+                if (executionHistory.getOverlappingOutputs() != null) {
+                    LOGGER.info("No leftover directories for {} will be deleted since overlapping outputs where detected.", task);
+                    cleanupDirectories = false;
+                }
+                taskOutputsGenerationListener.beforeTaskOutputsGenerated();
```

Comment:
Since we didn't call `taskArtifactState.afterTask` at the end - so we never took a new output snapshot and the files still counted as outputs of the old task. This is also why, while we still had overlapping outputs between Groovy and Java compilation, when moving source all files from `src/main/groovy` to `src/main/java` the `GroovyCompile` task would always remove the class from the Java compile task. With the new code, it should only remove the files the first time and then leave them there.